pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java

```diff
@@ -322,6 +322,7 @@ private void processResult(Record srcRecord,
                         srcRecord, t);
                 stats.incrUserExceptions(t);
                 srcRecord.fail();
+                throw new RuntimeException(t);
```

Comment:
The error here is an exception that happened in the user function not sink.  Also because it is in the result.whenComplete clause, anything thrown in there will not be bubbled to the main thread

Reply:
The call to the Sink write method is wrapped in a try/catch block that re-throws any exception it encounters. 
https://github.com/apache/pulsar/blob/7a89212afabd044c922a039c897df966557dff56/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java#L345